src/euroeval/benchmark_modules/litellm.py

```diff
@@ -452,44 +465,135 @@ def generate(self, inputs: dict) -> GenerativeModelOutput:
                 message=f"Failed to generate text, after {num_attempts} attempts."
             )
 
-        assert isinstance(model_response, ModelResponse)
-        if not model_response.choices:
-            # This happens for reasoning models, when they don't finish thinking and run
-            # out of tokens. Happens quite rarely, but we need to handle it.
+        model_output = self._create_model_output(model_responses=model_response)
+
+        return model_output
+
+    async def _generate_async(
+        self,
+        messages: list[dict[str, t.Any]],
+        generation_kwargs: dict[str, t.Any],
+        max_retries: int = 3,
+        max_reruns: int = 30,
+    ) -> tuple[list[ModelResponse], list[tuple[int, Exception]]]:
+        """Generate outputs from the model asynchronously.
+
+        Args:
+            messages (list[dict[str, t.Any]]):
+                The messages to pass to the model.
+            generation_kwargs (dict[str, t.Any]):
+                The generation kwargs to pass to the model.
+            max_retries (int):
+                The maximum number of retries to make.
+            max_reruns (int):
+                The maximum number of reruns to make.
+
+        Returns:
+            A tuple containing the successful responses and the failed responses.
+        """
+        success = []
+        failures = []
+
+        to_run = list(enumerate(messages))
+        for attempt in range(max_reruns + 1):
+            if not to_run:
+                break
+
+            requests = [
+                litellm.acompletion(
+                    messages=msg, max_retries=max_retries, **generation_kwargs
+                )
+                for _, msg in to_run
+            ]
+            responses = await asyncio.gather(*requests, return_exceptions=True)
+
+            next_to_run = []
+            for (orig_idx, _), response in zip(to_run, responses):
+                if isinstance(response, Exception):
+                    if attempt >= max_reruns:
+                        failures.append((orig_idx, response))
+                    else:
+                        next_to_run.append((orig_idx, messages[orig_idx]))
+                else:
+                    success.append(response)
+
+            to_run = next_to_run
+
+        return success, failures
+
+    def _create_model_output(
+        self, model_responses: list[ModelResponse]
+    ) -> GenerativeModelOutput:
+        """Create a GenerativeModelOutput object from a list of ModelResponse objects.
+
+        Args:
+            model_responses (list[ModelResponse]):
```

Comment:
Same as above, that we don't write types in docstrings:

```suggestion
            model_responses:
```